### DISCUSSION TITLE
How to use Font Awesome's otf fonts

### DISCUSSION THREAD

**Discussion by djeikyb** [AvaloniaUI/Avalonia#16736] [Q&A] [answered]
<details>
<summary>

Hiding most of the original post because of how wrong it is. tl;dr..

> I had a very, very hard time adding a font from Font Awesome. The [official guide][1] flatly did not work.

Because I misread and misunderstood the example code, repeatedly, in several ways. [Full explanation of where I went wrong.][0] Briefly, fonts are a special kind of embedded resource with their own nifty uri scheme. You shouldn't try fonts without understanding the basics of that uri scheme.

[0]: https://github.com/djeikyb/AvaloniaWithFontAwesome
[1]: https://docs.avaloniaui.net/docs/guides/styles-and-resources/how-to-use-fonts

</summary>

### 1. The problem

I had a very, very hard time adding a font from Font Awesome. The [official guide](https://docs.avaloniaui.net/docs/guides/styles-and-resources/how-to-use-fonts) flatly did not work. I don't know why. Maybe it's because my font file names have spaces. Maybe it's because the names inside the font file have spaces.

I was about to give up on custom fonts yet again when I found [this issue about Font Awesome not working.](https://github.com/AvaloniaUI/Avalonia/issues/13040) I'm also trying to use Font Awesome; great! There was some interesting information about otf fonts having multiple names. But there was no actual solution.

I decided to abandon the guide and look at how the Inter font is loaded. That looked pretty simple, if I could find the name of my font.

The rest of this post is probably filled with inaccuracies and is assuredly Not The Ideal Path. But, it _was_ a path that led out of the forest, eventually.

### 2. Finding the name of my font

I can't tell that there's any easy way to get this info with c#. I know that Avalonia does it, but I didn't see a readily available api. Python to the rescue!

[(adapted from jeremie bergeron's stackoverflow post)](https://stackoverflow.com/a/72228817/659715)

First some personal protective equipment:

```sh
python3 -m venv venv
source venv/bin/activate
```

A library called [fontTools](https://github.com/fonttools/fonttools) is doing the heavy lifting.

```
pip3 install fontTools
```

Now you can query for the name. Make sure to adjust the fontPath variable!

```python3
from fontTools import ttLib
fontPath='path/to/font.otf'
font = ttLib.TTFont(fontPath)
print(font['name'].getDebugName(16))
print(font['name'].getDebugName(1))
```

### 3. Following the working example of Inter

The code for [Avalonia.Fonts.Inter](https://github.com/AvaloniaUI/Avalonia/tree/11.1.3/src/Avalonia.Fonts.Inter) is pretty easy to follow.

I createad a new classlib project following the same naming convention. Create an Assets/ dir, mark them as AvaloniaResource in the csproj:

```xml
<ItemGroup>
  <AvaloniaResource Include="Assets\*" />
</ItemGroup>
```

The [AppBuilderExtension](https://github.com/AvaloniaUI/Avalonia/blob/11.1.3/src/Avalonia.Fonts.Inter/AppBuilderExtension.cs) is self explanatory. So is the [EmbeddedFontCollection](https://github.com/AvaloniaUI/Avalonia/blob/11.1.3/src/Avalonia.Fonts.Inter/InterFontCollection.cs). It doesn't seem to matter what the font uri is, as long as the syntax `font:some_name` is used.

```cs
public sealed class InterFontCollection : EmbeddedFontCollection
{
  public InterFontCollection() : base(
    new Uri("fonts:irrelevant", UriKind.Absolute), 
    new Uri("avares://Avalonia.Fonts.FontAwesome/Assets", UriKind.Absolute))
  {
  }
}
```

### 4. Use it

I picked a distinctive glyph to verify. You must use the name you got when querying font names for id 16. Probably if there was no name for id 16, then use id 1.

```xml
<TextBlock Text="$" FontSize="22" FontWeight="Light" FontFamily="Font Awesome 6 Sharp" />
```

</details>

**Comment by djeikyb** [comment]
And of course immediately after posting, I made a slight bit of headway with [the official guide's way.](https://docs.avaloniaui.net/docs/guides/styles-and-resources/how-to-use-fonts)

The guide suggests that
- the value of `x:Key` can be anything you want
- `#Nunito` must be the font name found at id 16

```xml
<Application.Resources>
  <FontFamily x:Key="NunitoFont">avares://GoogleFonts/Assets/Fonts#Nunito</FontFamily>
</Application.Resources>
```

This is not my experience. Instead
- the value of `x:Key` can be anything you want and you will never reference this value
- the hash and arg should be dropped entirely
- the value of a `TextElement.FontFamilyProperty` must be the font name found at id 16

I will guess that this is either a bug or a wild misunderstanding on my part. But it works?

```xml
<Application.Resources>
  <FontFamily x:Key="irrelevant">avares://Assets/Fonts</FontFamily>
</Application.Resources>
```

```xml
<TextBlock Text="$" FontSize="22" FontWeight="Light" FontFamily="Font Awesome 6 Sharp" />
```

**Reply by maxkatz6** [reply]
> the value of x:Key can be anything you want and you will never reference this value

Wrong, you actually need to reference it.
When you just write "Font Awesome 6 Sharp" - you might just remove font file from your assets :)

> the hash and arg should be dropped entirely
> the value of a TextElement.FontFamilyProperty must be the font name found at id 16

You are not using these font files, instead you let OS to find system font.

**Reply by djeikyb** [reply]
aw crap i didn’t think to test on a clean host. i’ll redact or delete soons i get a chance. i for sure have this font installed on the system

**Comment by maxkatz6** [comment, marked as answer]
Just tried to reproduce the issue, but couldn't:
1. mkdir font-test 
2. cd font-test 
3. dotnet new avalonia.app
4. Downloaded free desktop font awesome version from here https://fontawesome.com/download
5. Copied otf files to a new "Assets" folder (can be any folder, but just to be consistent with the guide)
6. Used first available tool to open font file - https://fontdrop.info/
7. Got that I want to display `F3A5` unicode character and my font name is `Font Awesome 6 Free`.
8. Added font to resources:
```xml
    <Application.Resources>
        <FontFamily x:Key="FontAwesome">avares://font-test/Assets/Font Awesome 6 Free-Regular-400.otf#Font Awesome 6 Free</FontFamily>
    </Application.Resources>
```
9. Created a text block using this font:
```xml
<TextBlock Text="&#xF3A5;" FontFamily="{StaticResource FontAwesome}" />
```

<img width="149" alt="Screenshot 2024-08-19 at 8 22 57 PM" src="https://github.com/user-attachments/assets/a2558d22-2e2a-465c-a28e-53be23ff0484">

**Reply by maxkatz6** [reply]
Also, a useful hint, font key lookup works via StartsWith, so "Font Awesome" key is also valid:
```xml
<FontFamily x:Key="FontAwesome">avares://font-test/Assets/Font Awesome 6 Free-Regular-400.otf#Font Awesome</FontFamily>
```

**Reply by djeikyb** [reply]
Ack! It was that i didn't realize the first part of the avares uri is the assembly name. That along with not understanding that the hash ref part is the font's Family Name / preferredFamily identifier. Thanks and wrote up what I hope is a much more correct guide around all the ways I shot myself in the foot: https://github.com/djeikyb/AvaloniaWithFontAwesome

**Comment by maxkatz6** [comment]
Also, there is nothing wrong with using EmbeddedFontCollection too. But that's more a solution if you want to ship your font as a nuget package. EmbeddedFontCollection single purpose there (and for Inter) is to keep font from resource being trimmed/removed from the assembly.

It doesn't really solve any other problem for a typical app usage.

[a comment by an automated account is not shown]
